pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/tenant/TenantRebalanceChecker.java

```diff
@@ -0,0 +1,307 @@
+        TenantRebalanceProgressStats progressStats =
+            JsonUtils.stringToObject(tenantRebalanceProgressStatsStr, TenantRebalanceProgressStats.class);
+        long statsUpdatedAt = Long.parseLong(jobZKMetadata.get(CommonConstants.ControllerJob.SUBMISSION_TIME_MS));
+
+        DefaultTenantRebalanceContext retryTenantRebalanceContext =
+            prepareRetryIfTenantRebalanceJobStuck(jobZKMetadata, tenantRebalanceContext, statsUpdatedAt);
+        if (retryTenantRebalanceContext != null) {
+          for (TenantRebalancer.TenantTableRebalanceJobContext ctx;
+              (ctx = retryTenantRebalanceContext.getOngoingJobsQueue().poll()) != null; ) {
+            abortTableRebalanceJob(ctx.getTableName());
+            // the existing table rebalance job is aborted, we need to run the rebalance job with a new job ID.
+            TenantRebalancer.TenantTableRebalanceJobContext newCtx =
+                new TenantRebalancer.TenantTableRebalanceJobContext(
+                    ctx.getTableName(), UUID.randomUUID().toString(), ctx.shouldRebalanceWithDowntime());
+            retryTenantRebalanceContext.getParallelQueue().addFirst(newCtx);
+          }
+          // the retry tenant rebalance job id has been created in ZK, we can safely mark the original job as
+          // aborted, so that this original job will not be picked up again in the future.
+          markTenantRebalanceJobAsAborted(jobId, jobZKMetadata, tenantRebalanceContext, progressStats);
+          retryTenantRebalanceJob(retryTenantRebalanceContext, progressStats);
+        } else {
+          LOGGER.info("Tenant rebalance job: {} is not stuck", jobId);
+        }
+      } catch (JsonProcessingException e) {
+        // If we cannot parse the job metadata, we skip this job
+        LOGGER.warn("Failed to parse tenant rebalance context for job: {}, skipping", jobId);
+      }
+    }
+  }
+
+  private void retryTenantRebalanceJob(DefaultTenantRebalanceContext tenantRebalanceContextForRetry,
+      TenantRebalanceProgressStats progressStats) {
+    ZkBasedTenantRebalanceObserver observer =
+        new ZkBasedTenantRebalanceObserver(tenantRebalanceContextForRetry.getJobId(),
+            tenantRebalanceContextForRetry.getConfig().getTenantName(),
+            progressStats, tenantRebalanceContextForRetry, _pinotHelixResourceManager);
+    ((DefaultTenantRebalancer) _tenantRebalancer).rebalanceWithContext(tenantRebalanceContextForRetry, observer);
+  }
+
+  /**
+   * Check if the tenant rebalance job is stuck, and prepare to retry it if necessary.
+   * A tenant rebalance job is considered stuck if:
+   * 1. There are no ongoing jobs, but there are jobs in the parallel or sequential queue, and the stats have not been
+   *    updated for longer than the heartbeat timeout.
+   * 2. There are ongoing table rebalance jobs, and at least one of them has not updated its status for longer than the
```

Comment:
I think these checks make sense but do add one line explaining the logic here (for instance, why we can't simply use heartbeat timeout check on the overall tenant rebalance stats). It's usually more useful to describe _why_ something is being done rather than simply describing _what_ is being done.